### ISSUE TITLE
Low performance of POD Attention compared to BatchPrefillWithPagedKVCache

### ISSUE THREAD

**Issue by Edenzzzz** [flashinfer-ai/flashinfer#1022]
Hi @yzh119 @AKKamath thanks for your effort in integrating this novel work into FlashInfer.
I recently made a PR to integrate POD Attn into SGLang, but found the performance is lower than BatchPrefillWithPagedKVCache. I ran a chunked prefill workload with input length 4000 and output length 200, feeding decode and prefill requests into the same kernel.
To support multiple prefill reqs I use a 2D mask to compute local attention in a packed sequence.
In profiling I see the POD Attn kernel always takes about 3 times longer to finish than the BatchPrefill kernel. 
<img width="1146" alt="Image" src="https://github.com/user-attachments/assets/db3f2725-f31c-49ac-b909-aece00fa46c1" />

<img width="826" alt="Image" src="https://github.com/user-attachments/assets/fdd61660-7f48-4724-8be6-dddb3b0aa8af" />

The results can be reproduced using [my branch](https://github.com/sgl-project/sglang/pull/5169)
```
export SGLANG_TORCH_PROFILER_DIR=/sgl-workspace/sglang/profile_log
python3 -m sglang.launch_server --model-path meta-llama/Meta-Llama-3.1-8B-Instruct --host 0.0.0.0 --disable-cuda-graph --enable-mixed-chunk --enable-pd-colocation
python3 -m sglang.bench_serving --backend sglang --dataset-name random --random-input 4000 --random-output 200 --request-rate 8 --num-prompt 480 --port 30000 --profile
```
Also sometimes [this line](https://github.com/flashinfer-ai/flashinfer/blob/9220fb3443b5a5d274f00ca5552f798e225239b7/include/flashinfer/attention/pod.cuh#L421) throws illegal memory access on kernel launch.
I wonder if you have any insights on why this might be the case? Thanks.

**Comment by Edenzzzz** [comment]
Looks like the `illegal memory access` is not due to shared memory, all the kernel calls before the error use the same smem, but the last one uses a larger `nblks` (2592)

<img width="741" alt="Image" src="https://github.com/user-attachments/assets/d54e03f5-776a-4109-b884-0d0b14d5edb2" />

**Comment by AKKamath** [comment]
I think I figured out the performance issue. I was going through the code in https://github.com/flashinfer-ai/flashinfer/blob/main/benchmarks/bench_mixed_attention.py for this.

The current POD Attention assumes a single input for prefill query, but right now a batch of prefills is being passed in.
For example in the first benchmark case 8 prefills each of size 17 are used. The corresponding KV is 8 of size 10000. Now this is being passed in as a single input, meaning the query length is 8 * 17 = 136, and the KV is 80000.
The attention operation then performs a Q * K of 136 * 80000. What we really wanted was 8 separate Q * K each of 17 * 10000. By this math, POD ends up doing 8 times more work than necessary, which is why we see very inflated latencies.

There are a few ways this could be fixed. The easiest but possibly least performant is to split the prefill and decode into batches and send one prefill and a set of decode per batch. Another option would be to add batch prefill support instead of single prefill. This will require quite a bit of code changes. The last option is to try to add POD support to the new Holistic kernel templates.

**Comment by Edenzzzz** [comment]
@AKKamath Could you try this code? I added a causal mask to hopefully remove the redundant work, but perf is still not as expected. I will look into the persistent and batch kernel

<img width="393" alt="Image" src="https://github.com/user-attachments/assets/f0e2a9c6-21b6-44ee-9e72-868aa11046d3" />

```
import numpy as np
import torch
from triton.testing import do_bench
import triton
import triton.language as tl
import flashinfer

MAX_KV_LEN =  16384  

@triton.jit
def create_causal_mask_from_page_triton(
    mask_ptr,  # [qo_len, kv_len]
    cu_qo_indptr,  # [bs + 1], cumulative ranges for each req
    cu_kv_lens_ptr,  # [bs + 1]
    prefix_lens_ptr,  # [bs + 1]
    stride_mask_qo,
    max_kv_len_per_req: tl.constexpr,  # must be static
):
    """
    Creates causal mask from paged indices to filter out kv for each request.
    Each program handles a row of qo,
    with the prefix kv attended by all queries in a request.
    """
    pid_qo = tl.program_id(axis=0)
    req_id = tl.program_id(axis=1)

    # Find which req this row belongs to
    qo_start = tl.load(cu_qo_indptr + req_id).to(tl.int32)
    qo_end = tl.load(cu_qo_indptr + req_id + 1).to(tl.int32)
    if qo_start <= pid_qo and pid_qo < qo_end:
        kv_start = tl.load(cu_kv_lens_ptr + req_id).to(tl.int32)
        kv_end = tl.load(cu_kv_lens_ptr + req_id + 1).to(tl.int32)
        kv_len_local = kv_end - kv_start
        # mask_offset = pid_qo * stride_mask_qo + kv_start + tl.arange(0, max_kv_len)

        kv_indices = tl.arange(0, max_kv_len_per_req)
        is_valid_kv = kv_indices < kv_len_local
        kv_pos = kv_indices
        qo_idx_local = (pid_qo - qo_start) + tl.load(prefix_lens_ptr + req_id + 1).to(
            tl.int32
        )
        mask = (kv_pos <= qo_idx_local) & is_valid_kv
        mask_offsets = pid_qo * stride_mask_qo + kv_pos + kv_start
        tl.store(mask_ptr + mask_offsets, mask, mask=is_valid_kv)


def run_bench(
    p_qo_lens,
    p_kv_lens,
    d_qo_lens,
    d_kv_lens,
    # page_block_size=1,
    num_kv_heads=4,
    num_qo_heads=28,
    head_dim=128,
    device=0,
    causal=True,
):
    # POD Attention only supports page size = 1 due to use of single prefill kernel
    page_block_size = 1
    seq_lens = torch.tensor(d_kv_lens + p_kv_lens, dtype=torch.int32)
    q_lens = torch.tensor(d_qo_lens + p_qo_lens, dtype=torch.int32)

    seq_lens_blocks = torch.ceil(seq_lens / page_block_size).int()
    d_seq_lens_blocks = (
        torch.tensor(d_kv_lens, dtype=torch.int32) / page_block_size
    ).int()

    q_indptr = torch.cat([torch.tensor([0]), torch.cumsum(q_lens, 0)], dim=0).int()
    kv_indptr = torch.cat(
        [torch.tensor([0]), torch.cumsum(seq_lens_blocks, 0)], dim=0
    ).int()
    d_q_indptr = torch.cat(
        [torch.tensor([0]), torch.cumsum(torch.tensor(d_qo_lens), 0)], dim=0
    ).int()
    d_kv_indptr = torch.cat(
        [torch.tensor([0]), torch.cumsum(d_seq_lens_blocks, 0)], dim=0
    ).int()
    num_blocks = kv_indptr[-1].item()

    q = torch.rand(q_indptr[-1].item(), num_qo_heads, head_dim).to(
        device, dtype=torch.bfloat16
    )
    kv_data = torch.randn(num_blocks, 2, page_block_size, num_kv_heads, head_dim).to(
        device, dtype=torch.bfloat16
    )

    workspace_buffer = torch.empty(128 * 1024 * 1024, dtype=torch.uint8, device=device)
    kv_layout = "NHD"

    wrapper_old = flashinfer.BatchPrefillWithPagedKVCacheWrapper(
        workspace_buffer,
        kv_layout=kv_layout,
        backend="fa2",
    )
    last_page_len = (seq_lens - 1) % page_block_size + 1
    wrapper_old.plan(
        q_indptr.to(device),
        kv_indptr.to(device),
        torch.arange(num_blocks).int().to(device),
        last_page_len,
        num_qo_heads,
        num_kv_heads,
        head_dim,
        page_block_size,
        causal=causal,
        q_data_type=torch.bfloat16,
        kv_data_type=torch.bfloat16,
    )
    ms_old = do_bench(lambda: wrapper_old.run(q, kv_data))

    if len(p_kv_lens) > 0:
        q_d = q[: d_q_indptr[-1]]
        kv_d = kv_data[: d_kv_indptr[-1]].unbind(1)
        q_p = q[d_q_indptr[-1] :]
        k_p, v_p = kv_data[d_kv_indptr[-1] :].unbind(1)
        k_p, v_p = k_p.squeeze(1), v_p.squeeze(1)
        
        p_q_lens = torch.tensor(p_qo_lens, dtype=torch.int32, device=device)
        p_seq_lens_blocks = torch.tensor(p_kv_lens, dtype=torch.int32, device=device) 
        p_q_indptr = torch.cat( 
            [torch.tensor([0], device=device), torch.cumsum(p_q_lens, 0)], dim=0, 
        ).int()
        p_kv_indptr = torch.cat(
            [torch.tensor([0], device=device), torch.cumsum(p_seq_lens_blocks, 0)], dim=0
        ).int()
        # Per-request mask for prefill
        mask = torch.zeros((p_q_indptr[-1], p_kv_indptr[-1]), dtype=torch.bool, device=device)
        
        bs_p = len(p_q_lens)
        max_kv_len = max(p_kv_lens)
        global MAX_KV_LEN
        MAX_KV_LEN = MAX_KV_LEN if max_kv_len <= MAX_KV_LEN else max_kv_len
        create_causal_mask_from_page_triton[(p_q_indptr[-1], bs_p)](
            mask,
            p_q_indptr,
            p_kv_indptr,
            p_kv_indptr,
            mask.stride(0),
            MAX_KV_LEN
        )
        
        last_page_len_d = (d_seq_lens_blocks - 1) % page_block_size + 1
        wrapper_pod = flashinfer.PODWithPagedKVCacheWrapper(
            workspace_buffer,
            kv_layout=kv_layout,
        )
        wrapper_pod.plan(
            d_q_indptr.to(device),
            d_kv_indptr.to(device),
            last_page_len=last_page_len_d,
            num_qo_heads=num_qo_heads,
            num_kv_heads=num_kv_heads,
            head_dim=head_dim,
            page_size=page_block_size,
            q_data_type=torch.bfloat16,
            kv_data_type=torch.bfloat16,
        )
        ms_pod = do_bench(
            lambda: wrapper_pod.run(
                q_p,
                k_p,
                v_p,
                q_d,
                kv_d,
                causal_p=causal,
                causal_d=causal,
                custom_mask_p=mask,
            )
        )

    print(f"Elapsed time (Batched Prefill): {ms_old:.2f} ms")
    if len(p_kv_lens) > 0:
        print(f"Elapsed time (POD Attention): {ms_pod:.2f} ms")
    total_bytes = (
        q.numel() * q.element_size() + kv_data.numel() * kv_data.element_size()
    )
    print(f"Loading memory size (MB): {total_bytes / (1024**2):.2f} MB")

    bandwidth_old_gb_s = total_bytes / (ms_old * 1e-3) / (1024**3)

    print(f"Memory bandwidth (Batched Prefill): {bandwidth_old_gb_s:.2f} GB/s")
    if len(p_kv_lens) > 0:
        bandwidth_pod_gb_s = total_bytes / (ms_pod * 1e-3) / (1024**3)
        print(f"Memory bandwidth (POD Attention): {bandwidth_pod_gb_s:.2f} GB/s")


if __name__ == "__main__":
    np.random.seed(42)
    torch.random.manual_seed(42)

    # Irregular sequence lengths for prefill and decode
    d_q_len_configs = [[1] * 122, [1] * 128, [1] * 242, [1] * 256]
    d_kv_len_configs = [[600] * 122, [10000] * 128, [400] * 242, [8192] * 256]
    p_q_configs = [[17] * 8, [], [17] * 16, []]
    p_kv_configs = [[10000] * 8, [], [8192] * 16, []]

    # construct random length testcases
    for _ in range(1):
        bsz = 256
        stride = 16
        sparsity = 0.05

        full_kv_len = np.random.randint(1000, 8192, size=bsz)
        p_q_lens = []
        p_kv_lens = []
        d_q_lens = []
        d_kv_lens = []
        for i in range(bsz):
            if i % stride == 0:
                kv_len = full_kv_len[i]
                qo_len = stride + 1
                p_q_lens.append(qo_len)
                p_kv_lens.append(kv_len)
            else:
                kv_len = int(full_kv_len[i] * sparsity)
                qo_len = 1
                d_q_lens.append(qo_len)
                d_kv_lens.append(kv_len)

        p_q_configs.append(p_q_lens)
        p_kv_configs.append(p_kv_lens)
        d_q_len_configs.append(d_q_lens)
        d_kv_len_configs.append(d_kv_lens)

    for _ in range(1):
        bsz = 128
        stride = 16
        sparsity = 0.05

        full_kv_len = np.random.randint(2000, 16000, size=bsz)
        p_q_lens = []
        p_kv_lens = []
        d_q_len = []
        d_kv_len = []

        for i in range(bsz):
            if i % stride == 0:
                kv_len = full_kv_len[i]
                qo_len = stride + 1
                p_q_lens.append(qo_len)
                p_kv_lens.append(kv_len)
            else:
                kv_len = int(full_kv_len[i] * sparsity)
                qo_len = 1
                d_q_lens.append(qo_len)
                d_kv_lens.append(kv_len)

        p_q_configs.append(p_q_lens)
        p_kv_configs.append(p_kv_lens)
        d_q_len_configs.append(d_q_lens)
        d_kv_len_configs.append(d_kv_lens)

    page_block_size = 1
    num_kv_heads = 4
    num_qo_heads = 28
    head_dim = 128

    for idx, (p_q_lens, p_kv_lens, d_q_len, d_kv_len) in enumerate(
        zip(p_q_configs, p_kv_configs, d_q_len_configs, d_kv_len_configs)
    ):

        print(f"===== Benchmark {idx+1}: (kv_len, qo_len) set =====")
        run_bench(
            p_q_lens,
            p_kv_lens,
            d_q_len,
            d_kv_len,
            # page_block_size=page_block_size,
            num_kv_heads=num_kv_heads,
            num_qo_heads=num_qo_heads,
            head_dim=head_dim,
            device=0,
            causal=True,
        )
```

**Comment by Edenzzzz** [comment]
Actually I think this doesn't avoid any redundant work😂, it still goes through all the kv

**Comment by AKKamath** [comment]
My goal is to add batch prefill support to POD. This should resolve all the issues. Unfortunately I have a few deadlines coming up, so this may take some time.

**Comment by Edenzzzz** [comment]
Sure thanks, I will also try working on that

**Comment by Edenzzzz** [comment]
When I changed **all the prefill batch sizes to 1**, there's still some slowdown

<img width="616" alt="Image" src="https://github.com/user-attachments/assets/2899b48b-9eb0-4c8a-b9b5-538980976bd3" />

**Comment by AKKamath** [comment]
This is the same benchmark, just all the p_q is 1 instead of 8/16?

**Comment by Edenzzzz** [comment]
Yeah I just did this. I'm using H100

<img width="663" alt="Image" src="https://github.com/user-attachments/assets/965d8d45-0e7a-4398-b250-9dc1361ebad9" />

**Comment by AKKamath** [comment]
Aha! I think you may have caught a bug. If you don't mind, could you printf the variable num_blocks_per_sm after Line 350 in include/flashinfer/attention/pod.cuh (here: https://github.com/flashinfer-ai/flashinfer/blob/68e1f2973a486f8ede79a718dd1f834cd624f51f/include/flashinfer/attention/pod.cuh#L351) then run the benchmark? The output should be 2 for an A100 GPU (maybe 1 for smaller GPUs). For some reason it's outputting 0 for me.

If this is the cause, I can quickly change it to manually calculate that variable instead of relying on the cudaOccupancyMaxActiveBlocksPerMultiprocessor call.

**Comment by Edenzzzz** [comment]
It's indeed outputting 0 on H100, I'm not sure why this is the case😂 maybe some illegal resource inputs?https://stackoverflow.com/questions/63114663/cuda-occupancy-max-active-blocks-per-multiprocessor-0
EDIT: Likely not, smem_size is just 64kb

<img width="466" alt="Image" src="https://github.com/user-attachments/assets/00a8a415-8505-4d4b-a842-e334cf9ecf38" />

<img width="536" alt="Image" src="https://github.com/user-attachments/assets/c64c5d81-4993-4b56-9689-962765fd2389" />

**Comment by AKKamath** [comment]
Yeah, I'm not exactly sure. Luckily, a fix should be pretty simple. Submitted here: https://github.com/flashinfer-ai/flashinfer/pull/1054

**Comment by AKKamath** [comment]
@Edenzzzz I remember you were trying to get POD-Attention working with a batch of prefills instead of a single one. Did you progress with that? I may have a bit of extra time now and be able to help get it working now.

**Comment by Edenzzzz** [comment]
@AKKamath I'm finishing up a prototype and will probably make a PR this weekend

**Comment by AKKamath** [comment]
Oh, amazing! Do let me know if you need any help

**Comment by Edenzzzz** [comment]
Sry got tied up a bit last weekend. Here's the branch and will upstream more tonight
https://github.com/Edenzzzz/flashinfer/tree/pod_batched

**Comment by yzh119** [comment]
Great! Looking forward to that!
